pkg/observer/controller_test.go

```diff
@@ -387,3 +389,378 @@ func Test_dropInProgressNodeGroups(t *testing.T) {
+}
+
+func TestPriority_FirstFailed_BlocksSecond(t *testing.T) {
+    scenario := test.BuildTestScenario(test.ScenarioOpts{Keys: []string{"a", "b"}, NodeCount: 1, PodCount: 1}).Flatten()
+    a := scenario.Nodegroups[0]
+    b := scenario.Nodegroups[1]
+    a.Spec.CycleSettings.Concurrency = 1
+    b.Spec.CycleSettings.Concurrency = 1
+    a.Spec.Priority = 0
+    b.Spec.Priority = 10
+
+    var objects []runtime.Object
+    objects = append(objects, a, b)
+
+    obs := testObserver{changed: map[string]*ListedNodeGroups{
+        a.Name: buildListed(a, scenario.Nodes[0].Name),
+        b.Name: buildListed(b, scenario.Nodes[0].Name),
+    }}
+    ctrl := newPriorityControllerForTest(t, objects, scenario.Nodes, obs)
+
+    // Create A
+    ctrl.Run()
+    lst1, _ := generation.ListCNRs(ctrl.client, &client.ListOptions{Namespace: ctrl.Namespace})
+    assert.Equal(t, 1, len(lst1.Items))
+
+    // Mark A as Failed
+    var cnrA atlassianv1.CycleNodeRequest
+    _ = ctrl.client.Get(context.TODO(), client.ObjectKey{Namespace: ctrl.Namespace, Name: lst1.Items[0].Name}, &cnrA)
+    cnrA.Status.Phase = atlassianv1.CycleNodeRequestFailed
+    _ = ctrl.client.Update(context.TODO(), &cnrA)
+
+    // Subsequent runs should never create B while A is Failed (treated as in-progress)
+    ctrl.Run()
+    ctrl.Run()
+    lst2, _ := generation.ListCNRs(ctrl.client, &client.ListOptions{Namespace: ctrl.Namespace})
+    assert.Equal(t, 1, len(lst2.Items))
+}
+
+func TestPriority_DefaultingNegativeAsZero_Sequential(t *testing.T) {
+    scenario := test.BuildTestScenario(test.ScenarioOpts{Keys: []string{"a", "b"}, NodeCount: 1, PodCount: 1}).Flatten()
+    a := scenario.Nodegroups[0]
+    b := scenario.Nodegroups[1]
+    a.Spec.CycleSettings.Concurrency = 1
+    b.Spec.CycleSettings.Concurrency = 1
+    a.Spec.Priority = -5 // should default to 0
+    b.Spec.Priority = 1
+
+    var objects []runtime.Object
+    objects = append(objects, a, b)
+
+    obs := testObserver{changed: map[string]*ListedNodeGroups{
+        a.Name: buildListed(a, scenario.Nodes[0].Name),
+        b.Name: buildListed(b, scenario.Nodes[0].Name),
+    }}
+    ctrl := newPriorityControllerForTest(t, objects, scenario.Nodes, obs)
+
+    // Run 1: only A should be created
+    ctrl.Run()
+    lst1, _ := generation.ListCNRs(ctrl.client, &client.ListOptions{Namespace: ctrl.Namespace})
+    assert.Equal(t, 1, len(lst1.Items))
+
+    // Mark A Successful, then B should be created
+    var cnrA atlassianv1.CycleNodeRequest
+    _ = ctrl.client.Get(context.TODO(), client.ObjectKey{Namespace: ctrl.Namespace, Name: lst1.Items[0].Name}, &cnrA)
+    cnrA.Status.Phase = atlassianv1.CycleNodeRequestSuccessful
+    _ = ctrl.client.Update(context.TODO(), &cnrA)
+
+    ctrl.Run()
+    lst2, _ := generation.ListCNRs(ctrl.client, &client.ListOptions{Namespace: ctrl.Namespace})
+    assert.Equal(t, 2, len(lst2.Items))
+}
+
+func TestPriority_ThreeLevels_Sequential(t *testing.T) {
```

Comment:
I'd like to see one more complex test case, which is that if changes keep coming in for a lower priority node group then Cyclops will keep putting off the higher priority node group from cycling.

You can reuse the scaffolding from this test to start it, and make it do something like this:
1. A is created
2. A succeeds
3. B is created
4. New changes come in for A
5. A is created
6. B succeeds
7. C is NOT created yet
8. A succeeds
9. C is now created
10. C succeeds